### ISSUE TITLE
Tensorflow python version incompatibility

### ISSUE THREAD

**Issue by icebreaker2** [tensorflow/tensorflow#19775]
Hey,

I use `inceptionv3` to classify various images. I experienced that images are classified with a different probabillity on different versions of python. I set up thre environments:
1. Tensorflow (b'v1.8.0-0-g93bc2e2072' 1.8.0) with python 3.6 with anaconda (no gpu support) (Windows 10)
2. Tensorflow (b'v1.8.0-0-g93bc2e2072' 1.8.0) with python 2.7.12 installed via plain pip. (Ubuntu 16.04)
3, Tensorflow (b'v1.8.0-0-g93bc2e2072' 1.8.0) with python 3.5.2 installed via plain pip. (Ubuntu 16.04)

Environment 1 and 3 give the same results. Using python2 gives the lower probabilities on the correct class than given by the other two setups.

Heres a little example I just created to outline the error and show how the classification performed:
```python
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image
import json, os

import tensorflow as tf
import tensorflow.contrib.slim as slim
from tensorflow.contrib.slim.nets import inception

FLAGS = tf.flags.FLAGS
tf.flags.DEFINE_string('image_name', 'cropped_panda.jpg', 'Image name with file ending.')
tf.flags.DEFINE_integer('top_k', 5, 'Top k predictions to print out.')
tf.flags.DEFINE_integer('image_height', 299, 'The image height for inception_v3 model.')
tf.flags.DEFINE_integer('image_width', 299, 'The image width for inception_v3 model.')
tf.flags.DEFINE_integer('image_channels', 3, 'The number of channels of for inception_v3 model.')
tf.flags.DEFINE_integer('image_N_classes', 1000, 'The number of classes in the inception_v3 model.')
tf.flags.DEFINE_float('weight_decay', 0.0, 'The weight decay applied to each edge/weight at each update step.')


def load_images(batch_shape, pathPrefix='../../images/'):
    """Retrieve numpy arrays of images and labels, read from a directory."""
    num_images = batch_shape[0]
    image_path = os.path.join(pathPrefix, FLAGS.image_name)
    image = Image.open(image_path)

    # crop
    wide = image.width > image.height
    new_w = FLAGS.image_width if not wide else int(image.width * FLAGS.image_width / image.height)
    new_h = FLAGS.image_height if wide else int(image.height * FLAGS.image_height / image.width)

    images = np.zeros(batch_shape)
    for idx in range(num_images):
        image = image.resize((new_w, new_h)).crop((0, 0, FLAGS.image_width, FLAGS.image_height))
        image = (np.asarray(image) / 255.0).astype(np.float32)
        images[idx, :, :, :] = image
    return images


def get_logits(image_placeholder):
    preprocessed = tf.multiply(tf.subtract(image_placeholder, 0.5), 2.0)
    arg_scope = inception.inception_v3_arg_scope(weight_decay=FLAGS.weight_decay)
    with slim.arg_scope(arg_scope):
        logits, _ = inception.inception_v3(
            preprocessed, FLAGS.image_N_classes + 1, is_training=False, reuse=False)
        logits = logits[:, 1:]
        probs = tf.nn.softmax(logits)
    return logits, probs


def get_inception_session(sess=None, pathToInception='../../'):
    saver = tf.train.Saver()
    saver.restore(sess, os.path.join(pathToInception, 'inception_v3.ckpt'))
    return sess


def main(argsunused):
    images = load_images((1, FLAGS.image_height, FLAGS.image_width, FLAGS.image_channels))
    image_placeholder = tf.placeholder(tf.float32, (1, FLAGS.image_height, FLAGS.image_width, FLAGS.image_channels))
    logits, preds = get_logits(image_placeholder)

    with tf.Session() as sess:
        sess = get_inception_session(sess)
        top_k = tf.nn.top_k(preds, k=FLAGS.top_k)
        values, indices = sess.run([top_k.values, top_k.indices], feed_dict={image_placeholder: images})
        for (index, value) in zip(indices[0], values[0]):
            print("Class %d with prob: %f" % (index, value))


if __name__ == '__main__':
    tf.app.run()
```
If I misssused something here, please be so kind and explain my mistake:)

Regards

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by icebreaker2** [comment]
Thanks for the response. Just added the missing fields.

**Comment by reedwm** [comment]
@icebreaker2, what probabilities do you get? Do you experience this issue when using Imagenet on the pretrained models [here](https://github.com/tensorflow/models/tree/master/research/slim)?

/CC @sguada any ideas why inception would perform differently on Python 2 vs 3?

**Comment by icebreaker2** [comment]
In the following I added the top5 results of the common giant panda example with inceptionv3.

With Python 2 and inceptionv3 I get (panda example, top5):
```shell
Class 388 with prob: 0.816696
Class 384 with prob: 0.010089
Class 812 with prob: 0.004649
Class 387 with prob: 0.002653
Class 995 with prob: 0.001971
```

With Python 3 and inceptionv3 I get (panda example, top5):
```shell
Class 388 with prob: 0.840522
Class 384 with prob: 0.008873
Class 812 with prob: 0.005004
Class 387 with prob: 0.003787
Class 995 with prob: 0.001669
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @sguada: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @sguada: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @sguada: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @sguada: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sguada** [comment]
I don't know why it would perform differently.

**Comment by reedwm** [comment]
/CC @gunan, any ideas why Python 2 and Python 3 would have numerical differences, leading to different results on Inception v3?

**Comment by gunan** [comment]
Python 2 and python 3 have quite a bit of differences. Here is a small number of them:
https://www.geeksforgeeks.org/important-differences-between-python-2-x-and-python-3-x-with-examples/

So it can really be as simple as division operator behaving differently, or maybe the two different python distros rounding numbers differently, or random number generators behaving differently. There are a lot of possibilities.

**Comment by reedwm** [comment]
Thanks for the info! I'm surprised there is a difference in accuracy between Python 2 and 3, because these differences should not affect the result of a TensorFlow graph. But, there is a difference, so marking as contributions welcome if anyone wants to investigate this.
